include/swift/SIL/SILInstruction.h

```diff
@@ -591,6 +710,123 @@ inline SingleValueInstruction *SILNode::castToSingleValueInstruction() {
            inst->getKind() <= SILInstructionKind::Last_##ID;    \
   }
 
+/// A value base result of a multiple value instruction.
+class MultipleValueInstructionResult : public ValueBase {
+  /// The parent instruction that this MultipleValueInstructionResult is
+  /// contained within.
+  MultipleValueInstruction *Parent;
+
+  /// The ownership kind assigned to this result by its parent.
+  ///
+  /// We store this rather than delegating to our parent since there is not a
+  /// natural clear way to derive the ValueOwnershipKind for all potential
+  /// multiple value instructions. This contrasts with the index of our result
+  /// in the parent array.
+  ValueOwnershipKind Kind;
+
+public:
+  MultipleValueInstructionResult(MultipleValueInstruction *parent, SILType type,
+                                 ValueOwnershipKind kind)
+      : ValueBase(ValueKind::MultipleValueInstructionResult, type),
+        Parent(parent), Kind(kind) {}
+
+  MultipleValueInstruction *getParent() const { return Parent; }
+
+  unsigned getIndex() const;
+
+  ValueOwnershipKind getOwnershipKind() const { return Kind; }
+
+  SILNode *getCanonicalSILNodeInObject();
+  const SILNode *getCanonicalSILNodeInObject() const;
+
+  static bool classof(const SILInstruction *) = delete;
+  static bool classof(const SILUndef *) = delete;
+  static bool classof(const SILNode *node) {
+    return node->getKind() == SILNodeKind::MultipleValueInstructionResult;
+  }
+};
+
+/// An instruction which always produces a fixed list of values.
+class MultipleValueInstruction : public SILInstruction {
+  static bool isMultipleValueInstKind(SILNodeKind kind) {
+    return kind >= SILNodeKind::First_MultipleValueInstruction &&
+           kind <= SILNodeKind::Last_MultipleValueInstruction;
+  }
+
+  friend class SILInstruction;
+  friend class SILInstructionResultArray;
+
+  /// An ArrayRef of Result objects. We assume that our child classes's will
+  /// allocate this for us on memory that will be placement newed.
+  ArrayRef<MultipleValueInstructionResult> Results;
+
+  SILInstructionResultArray getResultsImpl() const {
+    return SILInstructionResultArray(this);
+  }
+
+protected:
+  MultipleValueInstruction(
+      SILInstructionKind kind, SILDebugLocation loc,
+      ArrayRef<MultipleValueInstructionResult> results = {})
+      : SILInstruction(kind, loc), Results(results) {}
+
+  // Sometimes our subclasses need to set the new results /after/ we construct
+  // the MultipleValueInstruction.
+  void setResults(ArrayRef<MultipleValueInstructionResult> NewResults) {
+    Results = NewResults;
+  }
+
+public:
+  // Redeclare because lldb currently doesn't know about using-declarations
+  void dump() const;
```

Comment:
This is necessary in SingleValueInstruction because of multiple inheritance; it is not necessary here.